### PR TITLE
busybox: build additional applets and improve current applet features

### PR DESCRIPTION
**Pull request by jtru**
As [discussed on the OpenWrt forums](https://forum.openwrt.org/t/busybox-applets-included-in-openwrt-images-and-builds/165676/1) and briefly talked about with @hauke on IRC, I propose the following changes to the busybox default configuration for OpenWrt to provide more features for both interactive shell sessions and for scripting at the cost of a little binary size increase.

I hope the individual patches make it easy enough to cherry-pick what is deemed useful.

The busybox applets gained by merging this PR are the following:

-   comm
-   fold
-   nl
-   shuf
-   split
-   stat
-   tac
-   timeout
-   blkdiscard
-   ionice
-   mountpoint
-   nsenter
-   setpriv
-   unshare
-   inotifyd
-   tree
-   ts
-   watchdog
-   arping
-   pscan
-   fuser
-   nmeter
-   pkill
-   watch
-   chpst

Thank you for your consideration, and also for making OpenWrt happen! :sunglasses:

### CODE REVIEW

**Comment by mrkiko** [comment]
This should be conditional on LARGE flash devices in my opinion or, aniway, targets like ipq807x/amd64/i386 and so on.

**Comment by mrkiko** [comment]
Or - this might be better as a config option.

**Comment by jtru** [comment]
To clarify my intent: These *are* presently available configuration options already - but this is explicitly about affecting the *defaults*, because defaults do matter :)

To put the resulting binary size growth in perspective: For OpenWrt 21.02, the [recommended minimum flash size](https://openwrt.org/supported_devices/432_warning) was *16 MegaBytes*. This change will make busybox consume around *40 KiloBytes* (and that is the pessimistic end of the estimation) more.

I think that sacrifcing around *0.3% of total flash* for the multitude of features gained to use in scripts and on the shell is worth it, and not having these features/applets is not a sound compromise to make just to be able to drag along ancient hardware that has already been obsolete two releases ago still. Users with such constrained devices will have to build on their own and from source anyway, since they will have to compromise anyway, and can only choose in which particular way they are going to.

File: package/utils/busybox/Config-defaults.in [hunk 1]
```diff
@@ -24,7 +24,7 @@ config BUSYBOX_DEFAULT_FEATURE_VERBOSE_USAGE
 	default y
 config BUSYBOX_DEFAULT_FEATURE_COMPRESS_USAGE
 	bool
-	default n
+	default y
```
Comment (hauke) [inline package/utils/busybox/Config-defaults.in:27]: When the data is already compressed squashfs will not be able to compress it any more. This probably decreases the size of the busýbox binary, but it probably increases the size of the compressed root file system

Comment (jtru) [reply package/utils/busybox/Config-defaults.in:27]: That's certainly a valid concern - I was working with the ext4 image type for x86 using qemu to test my results, and had not considered that (even though I use squashfs on all my "production" OpenWrt devices...). I will try both variants - compressed and uncompressed busybox help texts - with a squashfs image, and report the image size results later today. Even a little bit of data should be better than no data to decide whether or not that's a worthwhile option :)

Comment (jtru) [reply package/utils/busybox/Config-defaults.in:27]: So I just built a defconfig squashfs images with all the proposed busybox changes (including compression of help texts) for x86/64/generic, and the resulting `openwrt-x86-64-generic-squashfs-rootfs.img.gz` measured *4098069* bytes.

Then, I repeat that build with identical configuration, except for *disabling* the `CONFIG_BUSYBOX_CONFIG_FEATURE_COMPRESS_USAGE` option, and the resulting `openwrt-x86-64-generic-squashfs-rootfs.img.gz` measured *4094607* bytes.

Ergo, enabling busybox help/usage text compression inside the gzipped squashfs image *increased* the size of that image by *3462* bytes.

**Comment by mrkiko** [comment]
On Sat, Jul 22, 2023 at 11:50:18AM -0700, Johannes Truschnigg wrote:
> To clarify my intent: These *are* presently available configuration options already - but this is explicitly about affecting the *defaults*, because defaults do matter :)
> 
> To put the resulting binary size growth in perspective: For OpenWrt 21.02, the [recommended minimum flash size](https://openwrt.org/supported_devices/432_warning) was *16 MegaBytes*. This change will make busybox consume around *40 KiloBytes* (and that is the pessimistic end of the estimation) more.
> 
> I think that sacrifcing around *0.3% of total flash* for the multitude of features gained to use in scripts and on the shell is worth it, and not having these features/applets is not a sound compromise to make just to be able to drag along ancient hardware that has already been obsolete two releases ago still. Users with such constrained devices will have to build on their own and from source anyway, since they will have to compromise anyway, and can only choose in which particular way they are going to.

Hi!

I don't see things this way. And in any case findthis change to be too "boad" in scope.
For example, I don't we'll need blkdiscard on a lot of 32M SPI flash devices.
In my opinion this should be limited to a target or a category of targets.

Thanks,
Enrico

**Comment by jtru** [comment]
Why don't you see things this way? :) Or, asked another way: How exactly do you see things?

> For example, I don't think we'll need blkdiscard on a lot of 32M SPI flash devices.

A similar argument could be made for `hwclock`, which isn't needed on devices without an RTC, or `mkswap`/`swapon`, which isn't useful on devices without any additional storage - and probably others, too. (They are present in the default images regardless.)

I was trying to select additional applets that provide previously unavailable functionality by "unlocking" features from user space that are already supported in the kernel (such as `blkdiscard` and `unshare` would), and/or increase user comfort a great deal in my opinion and experience.


> And in any case find this change to be too "broad" in scope.

Do you have a concrete suggestion on what to change, and how much space that would save over my current proposal? I'd really appreciate that! :)

**Comment by mrkiko** [comment]
On Sun, Jul 23, 2023 at 03:02:30AM -0700, Johannes Truschnigg wrote:
> Why don't you see things this way? :) Or, asked another way: How exactly do you see things?

Sorry, I'll try to be more clar.

I think this change should be a config option, easier to turn on and off, depending on the environment you work on.
I think the change is perfpectly fine in spirit, but I think it should be not so broad in scope - the user should be able to enable or disable this, or at least it should be limited to some targets where storage is not a problem, like amd64/x86.
> 
> > For example, I don't think we'll need blkdiscard on a lot of 32M SPI flash devices.
> 
> A similar argument could be made for `hwclock`, which isn't needed on devices without an RTC, or `mkswap`/`swapon`, which isn't useful on devices without any additional storage - and probably others, too. (They are present in the default images regardless.)

Good point. Infact I would try to avoid making things worse.
> 
> I was trying to select additional applets that provide previously unavailable functionality by "unlocking" features from user space that are already supported in the kernel (such as `blkdiscard` and `unshare` would), and/or increase user comfort a great deal in my opinion and experience.

And probably a similar argument might be made for other kernel functionalities.
> 
> 
> > And in any case find this change to be too "broad" in scope.
> 
> Do you have a concrete suggestion on what to change, and how much space that would save over my current proposal? I'd really appreciate that! :)

I'd suggest to make this an option and/or to make this target dependant, so that the extra space will be used when it is less of a problem.
Making this an option would also allow users to opt-in or out more easily.

Thanks,
Enrico

**Comment by jtru** [comment]
Thanks for elaborating, I think I do understand your concerns better now :) But I think the situation is less dire than you might assume - I tried to find devices supported by 22.03.5 today (that is, having an official image available via the Firmware Selector), and could not find even a single device that would not be able to support the proposed/incurred increase in image size.

As I said, all the required knobs to *undo* the effects of this patch are already present, as this only is about changing the **defaults** for generated OpenWrt images - users who compile themselves just need to change a few `Y`s to `N`s in their `.config`, and everything is to them as it was before. And users/device owners who would actually benefit from saving these ~20K by not having these busybox utils available *will have to configure and compile their images themselves anyway*.

Btw, I don't think making this a target-dependent change is technically possible (if it is, I am not aware how to do it).

I do actually think the addition of *some* of the proposed changes in this patch series is less warranted/desirable than others (having `blkdiscard` is a fringe interest for people who connect compatible storage to their devices, for example), but I would really like to add at the very least the stuff from commits 07ef4e7 and 58c522f. Don't see it as a waste of bytes - try to see it as an investment of bytes to enable useful features! ;)

**Comment by mrkiko** [comment]
On Sun, Jul 23, 2023 at 12:02:31PM -0700, Johannes Truschnigg wrote:
> Thanks for elaborating, I think I do understand your concerns better now :) But I think the situation is less dire than you might assume - I tried to find devices supported by 22.03.5 today (that is, having an official image available via the Firmware Selector), and could not find even a single device that would not be able to support the proposed/incurred increase in image size.
> 
> As I said, all the required knobs to *undo* the effects of this patch are already present, as this only is about changing the **defaults** for generated OpenWrt images - users who compile themselves just need to change a few `Y`s to `N`s in their `.config`, and everything is to them as it was before. And users/device owners who would actually benefit from saving these ~20K by not having these busybox utils available *will have to configure and compile their images themselves anyway*.

Ok, so please let this be undoable by unticking a single option, rather than having to go in the busybox config and having to search and untick the single options.

Thanks,
Enrico

**Comment by M95D** [comment]
Just my 2 cents:

@jtru 
> [...] to provide more features for both interactive shell sessions and for scripting [...]

> I think that sacrifcing around 0.3% of total flash for the multitude of features gained to use in scripts and on the shell is worth it.

You would be sabotaging 0.3% of the electronic waste reduction efforts of OpenWrt users and the "multitude of features" are not that many.

> Users with such constrained devices will have to build on their own and from source anyway, since they will have to compromise anyway, and can only choose in which particular way they are going to.

In my opinion, this argument is logically invalid. You can't expect devs to use these tools in scripts (as this is the stated motivation) and at the same time not break OpenWrt if they are removed by a poor user who bought a "constrained" device from the flea market.

OpenWrt is not a desktop OS. Shell is for admin tasks only, and most common admin tasks are done in Luci. So console tools have minimal use. But let me comment on each one:

> comm

Compare files with screen-formated output. That's not at all useful for scripts and 1) diff is a lot more capable for both scripts and console and 2) cmp is a lot simpler for "are they identical?" use-cases, probably faster too, if diff is too much.

> fold

Break lines in text? Why would anyone do that?!? This should be removed from all linuxes! I hate it when messages wrap at 78 columns on a 200-column screen and then force me to scroll vertically. [/rant]

> nl

Text formatting for script debugging. I think nobody does it directly on OpenWrt. Just use an editor; OpenWrt has lots of them. Or mount OpenWrt filesystem on your desktop and do it from there.

> shuf

Load balancing by randomising server list? Yes, this one may actually be useful in some rare cases.

> split

OK. If anyone wants to make OpenWrt a storage server, this may be of use.

> stat and nmeter

There are live stats and statistics with very nice graphics in Luci, htop, atop and other *tops in console, and there's collectd and snmp for remote logging. Why stat?

> tac

Use cat more wisely?

> timeout

Yes. This should be included.

> blkdiscard

There are no SSDs on routers and USB adapters probably don't support discard. If anyone wants to run OpenWrt on their desktop machine and do manual management of their SSDs, just add blkdiscard manually. They are probably going to build their own customized OpenWrt anyway for stuff like keyboards and display drivers. There's a blkdiscard package too.

> ionice

For an OpenWrt storage server that runs a database simultaneously? I have it right now and it works fine without ionice, but it would be *nice*. :-) Oh, wait, there's a *nice* package with ionice! :-)

If you want it for scripts, note that CPU nice for services is done with procd.

> mountpoint

YES! Double YES! This should be added!

> nsenter and unshare

AFAIK, the kernel doesn't have namespaces by default.

> setpriv and chpst

I never used these. Isn't seccomp better?

> inotifyd

I use audit.

> tree

Yes. It would be useful for storage servers. Listing trees via nfs is sooo slooow...

> ts

Very useful for programs that don't do logs properly, but most programs do and scripts should use *date* instead.

> watchdog

I think there is an OpenWrt watchdog already.

> arping

No core service uses it (what for?) and there's a package for those who need it.

> pscan

For those who forget what port they moved ssh to? Use /etc/services. Or maybe you want an OpenWrt honeypot? Most users wouldn't want to make life easy for hackers, so no pscan.

> fuser

lsof | grep and audit.

> pkill

init.d scripts are not good enough?

> watch

This would be nice to have by default, but it's easy to use an infinite loop with reset and sleep, maybe an alias for it. Again, you don't need to watch the screen waiting for something to happen; audit and collectd will save your time.

@mrkiko 
> Ok, so please let this be undoable by unticking a single option, rather than having to go in the busybox config and having to search and untick the single options.

Please, no! Busybox already has lots of config options, adding dependecies like this will make it even more difficult to config. Just imagine searching 5 minutes for an option only to find it preselected and unable to turn it off. I would go nuts!

Thanks for reading.

**Pull request closed without merging**